File: app/src/main/java/org/kiwix/kiwixmobile/di/modules/ApplicationModule.java
/*
 * Kiwix Android
 * Copyright (C) 2018  Kiwix <android.kiwix.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.kiwix.kiwixmobile.di.modules;

import android.app.Application;
import android.app.DownloadManager;
import android.app.NotificationManager;
import android.content.Context;
import android.location.LocationManager;
import android.os.storage.StorageManager;
import dagger.Module;
import dagger.Provides;
import dagger.android.AndroidInjectionModule;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import javax.inject.Singleton;
import org.kiwix.kiwixmobile.di.qualifiers.Computation;
import org.kiwix.kiwixmobile.di.qualifiers.IO;
import org.kiwix.kiwixmobile.di.qualifiers.MainThread;
import org.kiwix.kiwixmobile.downloader.DownloadMonitor;
import org.kiwix.kiwixmobile.downloader.fetch.FetchDownloadMonitor;
import org.kiwix.kiwixmobile.utils.BookUtils;

@Module(includes = {
  ActivityBindingModule.class,
  AndroidInjectionModule.class,
  DownloaderModule.class,
  ViewModelModule.class,
  DatabaseModule.class
})
public class ApplicationModule {

  @Provides @Singleton Application provideApplication(Context context) {
    return (Application) context;
  }

  @Provides
  @Singleton
  NotificationManager provideNotificationManager(Context context) {
    return (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
  }

  @Provides @Singleton DownloadManager provideDownloadManager(Context context) {
    return (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
  }

  @Provides
  @Singleton
  BookUtils provideBookUtils() {
    return new BookUtils();
  }

  @IO
  @Provides
  public Scheduler provideIoThread() {
    return Schedulers.io();
  }

  @MainThread
  @Provides
  public Scheduler provideMainThread() {
    return AndroidSchedulers.mainThread();
  }

  @Computation
  @Provides
  public Scheduler provideComputationThread() {
    return Schedulers.computation();
  }

  @Provides @Singleton
  LocationManager provideLocationManager(Context context) {
    return (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
  }

  @Provides @Singleton
  DownloadMonitor provideDownloadMonitor(FetchDownloadMonitor fetchDownloadMonitor) {
    return fetchDownloadMonitor;
  }

  @Provides @Singleton
  StorageManager provideStorageManager(Context context) {
    return (StorageManager) context.getSystemService(Context.STORAGE_SERVICE);
  }
}

File: core/src/main/java/org/kiwix/kiwixmobile/core/di/modules/ApplicationModule.kt
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile.core.di.modules

import android.app.Application
import android.app.NotificationManager
import android.content.Context
import android.net.ConnectivityManager
import android.net.wifi.WifiManager
import android.os.storage.StorageManager
import dagger.Module
import dagger.Provides
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import org.kiwix.kiwixmobile.core.DarkModeConfig
import org.kiwix.kiwixmobile.core.data.remote.ObjectBoxToLibkiwixMigrator
import org.kiwix.kiwixmobile.core.data.remote.ObjectBoxToRoomMigrator
import org.kiwix.kiwixmobile.core.di.qualifiers.Computation
import org.kiwix.kiwixmobile.core.di.qualifiers.IO
import org.kiwix.kiwixmobile.core.di.qualifiers.MainThread
import org.kiwix.kiwixmobile.core.downloader.DownloadMonitor
import org.kiwix.kiwixmobile.core.downloader.downloadManager.DownloadManagerMonitor
import org.kiwix.kiwixmobile.core.reader.ZimFileReader
import org.kiwix.kiwixmobile.core.utils.BookUtils
import javax.inject.Singleton

@Module(
  includes = [
    DownloaderModule::class,
    DatabaseModule::class
  ]
)
class ApplicationModule {

  @Provides
  @Singleton
  internal fun provideApplication(context: Context): Application =
    context as Application

  @Provides
  @Singleton
  internal fun provideNotificationManager(context: Context): NotificationManager =
    context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

  @Provides
  @Singleton
  internal fun provideBookUtils(): BookUtils = BookUtils()

  @Provides
  @Singleton
  fun provideObjectBoxToLibkiwixMigrator() = ObjectBoxToLibkiwixMigrator()

  @Provides
  @Singleton
  fun provideObjectBoxToRoomMigrator() = ObjectBoxToRoomMigrator()

  @IO
  @Provides
  fun provideIoThread(): Scheduler = Schedulers.io()

  @MainThread
  @Provides
  fun provideMainThread(): Scheduler = AndroidSchedulers.mainThread()

  @Computation
  @Provides
  fun provideComputationThread(): Scheduler = Schedulers.computation()

  @Provides
  @Singleton
  internal fun provideDownloadMonitor(
    downloadManagerMonitor: DownloadManagerMonitor
  ): DownloadMonitor = downloadManagerMonitor

  @Provides
  @Singleton
  internal fun provideStorageManager(context: Context): StorageManager =
    context.getSystemService(Context.STORAGE_SERVICE) as StorageManager

  @Provides
  @Singleton
  internal fun providesZimFileReaderFactory(darkModeConfig: DarkModeConfig):
    ZimFileReader.Factory = ZimFileReader.Factory.Impl(darkModeConfig)

  @Provides
  @Singleton
  fun provideConnectivityManager(context: Context): ConnectivityManager =
    context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

  @Provides
  @Singleton
  fun provideWifiManager(context: Context): WifiManager =
    context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
}
